pkg/model/autoscalinggroup.go

```diff
@@ -89,14 +89,47 @@ func (b *AutoscalingGroupModelBuilder) Build(c *fi.ModelBuilderContext) error {
 
 			{
 				associatePublicIP := true
-				if b.Cluster.IsTopologyPublic() {
+				switch ig.Spec.Role {
+				case kops.InstanceGroupRoleMaster:
+					switch b.Cluster.Spec.Topology.Masters {
+					case kops.TopologyPrivate:
+						associatePublicIP = false
+						// TODO: We probably should honor AssociatePublicIP
+
+					case kops.TopologyPublic:
+						associatePublicIP = true
+						if ig.Spec.AssociatePublicIP != nil {
+							associatePublicIP = *ig.Spec.AssociatePublicIP
+						}
+
+					default:
+						return fmt.Errorf("unhandled master topology %q", b.Cluster.Spec.Topology.Masters)
+					}
+
+				case kops.InstanceGroupRoleNode:
+					switch b.Cluster.Spec.Topology.Nodes {
+					case kops.TopologyPrivate:
+						associatePublicIP = false
+						// TODO: We probably should honor AssociatePublicIP
```

Comment:
You're right.  I changed the comment to `// TODO: what if AssociatePublicIP is set`.  We probably just need to reject this in validation.